extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/UnremovableBeanBuildItem.java

```diff
@@ -137,6 +139,24 @@ public static UnremovableBeanBuildItem beanTypes(Set<DotName> typeNames) {
         return new UnremovableBeanBuildItem(new BeanTypesExclusion(typeNames));
     }
 
+    /**
+     * Match beans whose bean types include one of the specified type names,
+     * be it raw or parameterized.
+     *
+     * @param rawTypeNames The raw types whose implementors should not be removed.
+     * @param indexView A Jandex index view allowing to list all implementors of the given raw types.
+     * @return a new build item
+     */
+    public static UnremovableBeanBuildItem beanTypesIncludingParameterized(Set<DotName> rawTypeNames, IndexView indexView) {
```

Comment:
I don't think this is necessary. For parameterized types we always compare the raw type, i.e. if you use `beanTypes(DotName.createSimple(ValueExtractor.class))` then if your bean implements `ValueExtractor<String>` it should be matched.

Reply:
Right... except that parameterized types using wildcards are ignored in compliance with the CDI spec, as you explained in https://github.com/quarkusio/quarkus/issues/36831#issuecomment-1794556282.

Which means `beanTypes(DotName.createSimple(ValueExtractor.class))` won't match a bean implementing `ValueExtractor<List<?>>` for example.

Which extensions may need something like this when they have no control over whether implementations will use wildcards. In the case of Hibernate Validator using wildcards in `ValueExtractor` implementations is even one of the recommended solutions, so... :/

Though this method should probably be renamed to translate this... ?

Reply:
Yes, the name should not contain `beanTypes` at all. 

Nevertheless, I believe that it's not a very common functionality so a custom `Predicate<BeanInfo>` in the HV extension would be a better fit.

Reply:
> Yes, the name should not contain beanTypes at all.

Sure. Though it's problematic that "bean type" is a very generic wording for a very specific thing in CDI :/ I suspect most people using `beanTypes(...)` methods, like me, didn't realize that CDI Bean types are not the same as the Java types of objects registered as CDI beans.

> Nevertheless, I believe that it's not a very common functionality

I already have a second extension needing this with Hibernate ORM and the `TenantResolver`, as you know: #36831

I just had a quick look at the current code in Quarkus, and I suspect we should be using it in other places as well:

* For `AttributeConverter<X, Y>`: what if a bean implements `AttributeConverter<List<? extends Number>, String>`? https://github.com/quarkusio/quarkus/blob/3fc255fde12523deb4a8127a0ab1950b35cfecd9/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L615
 * For `BeanMarshaller<T>`, `EnumMarshaller<E>`, `MessageMarshaller<T>`, `RawProtobufMarshaller<T>`: same idea. https://github.com/quarkusio/quarkus/blob/d7842f983f719c19785080da7179ce9c02c9e80a/extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java#L463
 * For `ServiceDiscoveryProvider<T>`, `LoadBalancerProvider<T>`, `ServiceRegistrarProvider<T, MetadataKeyType>`, `ServiceRegistrarLoader<MetadataKeyType>`: same idea. https://github.com/quarkusio/quarkus/blob/9bbfe72549539f5378929b6bc85e639d12c5da6e/extensions/smallrye-stork/deployment/src/main/java/io/quarkus/stork/deployment/SmallRyeStorkProcessor.java#L51-L57

Really, I'm pretty sure most extensions using `beanTypes(...)` were written without knowing that it wouldn't work with wildcard types. I.e., they **should** have been using something like this, and just didn't because they assumed `beanTypes(...)` would work intuitively, but it doesn't as soon as one uses wildcards, and nobody reported the problem yet.

So I agree applications ending up with a problem because of wildcards should be rare (except for Hibernate Validator). But I disagree this is an uncommon functionality, because pretty much every extension needing to register unremovable generic interfaces would need something like this, whether their authors know it or not.

Reply:
>  But I disagree this is an uncommon functionality, because pretty much every extension needing to register unremovable generic interfaces would need something like this, whether their authors know it or not.

Well, type variables are not a problem. And I'm not sure if wildcard type arguments are very common. But I agree that it's confusing for those who don't care about the CDI spec (i.e. majority of users ;-).

Reply:
Right, and by the way I'm not saying I (or extension maintainers, or application developers) don't care about the CDI spec. Just that they may not be aware of every edge case that it covers, especially those that are not particularly intuitive.

Now I'll try to be more constructive... let's continue the conversation there: https://github.com/quarkusio/quarkus/issues/36831#issuecomment-1794731890, where I suggested we should have a way to make CDI use the raw type as a bean type for beans that implement a parameterized type, but for specific raw types only of course. I suggested a build item, Matej suggested an annotation which extensions could apply through bytecode transformation, which I guess could be triggered by a build item as well. Could you please chime in there when you have some time?